cirq-core/cirq/optimizers/transformer_primitives.py

```diff
@@ -0,0 +1,186 @@
+    )
+
+
+def map_moments(
+    circuit: CIRCUIT_TYPE,
+    map_func: Callable[[ops.Moment, int], Sequence[ops.Moment]],
+) -> CIRCUIT_TYPE:
+    """Applies local transformation on moments, by calling `map_func(moment)` for each moment.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        map_func: Mapping function from (cirq.Moment, moment_index) to a sequence of moments.
+    """
+    return _to_target_circuit_type(
+        circuits.Circuit(map_func(circuit[i], i) for i in range(len(circuit))), circuit
+    )
+
+
+def map_operations(
+    circuit: CIRCUIT_TYPE,
+    map_func: Callable[[ops.Operation, int], ops.OP_TREE],
+) -> CIRCUIT_TYPE:
+    """Applies local transformations on operations, by calling `map_func(op)` for each op.
+
+    Note that the function assumes `issubset(qubit_set(map_func(op)), op.qubits)` is True.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        map_func: Mapping function from (cirq.Operation, moment_index) to a cirq.OP_TREE. If the
+            resulting optree spans more than 1 moment, it's inserted in-place in the same moment as
+            `cirq.CircuitOperation(cirq.FrozenCircuit(op_tree)).with_tags(MAPPED_CIRCUIT_OP_TAG)`
+            to preserve moment structure. Utility methods like `cirq.unroll_circuit_op` can
+            subsequently be used to unroll the mapped circuit operation.
+
+    Raises:
+          ValueError if `issubset(qubit_set(map_func(op)), op.qubits) is False`.
+    """
+
+    def apply_map(op: ops.Operation, idx: int) -> ops.OP_TREE:
+        c = circuits.FrozenCircuit(map_func(op, idx))
+        if not c.all_qubits().issubset(op.qubits):
+            raise ValueError(
+                f"Mapped operations {c.all_operations()} should act on a subset "
+                f"of qubits of the original operation {op}"
+            )
+        if len(c) == 1:
+            # All operations act in the same moment; so we don't need to wrap them in a circuit_op.
+            return c[0].operations
+        circuit_op = circuits.CircuitOperation(c).with_tags(MAPPED_CIRCUIT_OP_TAG)
+        return circuit_op
+
+    return map_moments(circuit, lambda m, i: [ops.Moment(apply_map(op, i) for op in m.operations)])
+
+
+def map_operations_and_unroll(
+    circuit: CIRCUIT_TYPE,
+    map_func: Callable[[ops.Operation, int], ops.OP_TREE],
+) -> CIRCUIT_TYPE:
+    """Applies local transformations via `cirq.map_operations` & unrolls intermediate circuit ops.
+
+    See `cirq.map_operations` and `cirq.unroll_circuit_op` for more details.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        map_func: Mapping function from (cirq.Operation, moment_index) to a cirq.OP_TREE.
+    """
+    return unroll_circuit_op(map_operations(circuit, map_func))
```

Comment:
Is this really needed ?

Reply:
The `map_operations_and_unroll` method? It's expected to be the most common use case of these two primitives and hence the convenience method will reduce 2 function calls to 1 function call at all call sites.